### PR TITLE
ULL-IACTEC Integration of USB Infrared & Ultraviolet Cameras

### PR DESCRIPTION
**Pull request by fmarcanoull**
Integration of Thermal Expert Infrared & PCO Ultraviolet USB Cameras. First version. Project developed by University of La Laguna & IAC Medical Technologies.

### CODE REVIEW

File: src/PlusDataCollection/CMakeLists.txt [hunk 1]
```diff
@@ -2314,6 +2316,95 @@ IF(PLUS_USE_INFRARED_TEQ1_CAM)
...
+    ${PCOUV_SDK_LIBRARY}
+    ${PCOUV_SDK_INCLUDE_DIR}
+  )
+
+  LIST(APPEND ${PROJECT_NAME}_LIBS
+    ${PCOUV_SDK_LIBRARY}
+  )
+  LIST(APPEND External_Libraries_Install
+    ${PCOUV_SDK_BINARY}/${SDK_FILENAME}
+    #${PCOUV_SDK_BINARY}/${SDK_FILENAME_USB}
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/CMakeLists.txt:2394]: Remove unused code.

File: src/PlusDataCollection/CMakeLists.txt [hunk 2]
```diff
@@ -2314,6 +2316,95 @@ IF(PLUS_USE_INFRARED_TEQ1_CAM)
...
+    #${PCOUV_SDK_BINARY}/${SDK_FILENAME_IMG}
+  )
+  LIST(APPEND External_Libraries_Debug
+    ${PCOUV_SDK_BINARY}/${SDK_FILENAME}
+    #${PCOUV_SDK_BINARY}/${SDK_FILENAME_USB}
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/CMakeLists.txt:2399]: Remove unused code.

File: src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  }
+  return PLUS_SUCCESS;
+}
+
+//----------------------------------------------------------------------------
+PlusStatus vtkInfraredTEEV2Cam::InternalConnect()
+{
+    SCANINFO* pScan = new SCANINFO[MAX_USB_NUM];
+    // FM MODIFICACION 2023-01-08. En vtkInfraredTEEV2Cam.cxx, eliminaci�n de primer par�metro de ScanTE
+    // ScanTE(nullptr, pScan);
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx:84]: Remove unused code.

File: src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    {
+        LOG_ERROR("Thermal Expert EV2: camera not detected.");
+        return PLUS_FAIL;
+    }
+    else{
+        LOG_DEBUG("Thermal Expert EV2 Camera handle: " << this->device);
+
+    }
+  // FM MODIFICACION 2023-01-08. En vtkInfraredTEEV2Cam.cxx
+  // this->pTE = OpenTE_B(nullptr, I3_TE_EV2, this->device);
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx:107]: Remove unused code.

File: src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  this->pTE = OpenTE_A(this->device);
+
+  if (! this->pTE)
+  {
+    LOG_ERROR("Thermal Expert EV2: Failed to open.");
+    return PLUS_FAIL;
+  }
+
+  // FM MODIFICACION 2023-01-08. En los ejemplos de adquisici�n de la c�mara, no existe un m�todo similar a ReadFlashData. Se elimina su uso.
+  //if (this->pTE->ReadFlashData() == 1)
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx:117]: Remove unused code.

File: src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+  }
+
+  if (!this->pTE->RecvImage(this->pImgBuf))
+  {
+    LOG_ERROR("vtkInfraredTEEV2Cam::InternalUpdate Unable to receive frame");
+    return PLUS_SUCCESS; // VERIFICAR ESTO
+  }
+
+  // FM MODIFICACION 2023-01-08. En vtkInfraredTEEV2Cam.cxx, eliminaci�n de  de CalcEntireTemp
+  // pTE->CalcEntireTemp(this->pImgBuf, false);
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx:157]: Remove unused code.

File: src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    // Init the buffer with the metadata from the first frame
+    aSource->SetImageType(US_IMG_BRIGHTNESS);
+    aSource->SetPixelType(VTK_UNSIGNED_SHORT); // FM MODIFICACION 2023-01-09, para tipo EV2: aSource->SetPixelType(VTK_TYPE_FLOAT32);    
+    aSource->SetNumberOfScalarComponents(1);
+    aSource->SetInputFrameSize(this->width, this->height, 1);
+  }
+
+  // Add the frame to the stream buffer
+  FrameSizeType frameSize = { static_cast<unsigned int>(this->width), static_cast<unsigned int>(this->height), 1 };
+  // FM MODIFICACION 2023-01-09, para tipo EV2: if (aSource->AddItem(this->pImgBuf, aSource->GetInputImageOrientation(), frameSize, VTK_TYPE_FLOAT32, 1, US_IMG_BRIGHTNESS, 0, this->FrameNumber) == PLUS_FAIL)
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/InfraredTEEV2Cam/vtkInfraredTEEV2Cam.cxx:177]: Remove unused code.

File: src/PlusDataCollection/Testing/CMakeLists.txt [hunk 8]
```diff
@@ -507,6 +507,28 @@ IF(PLUS_USE_INFRARED_TEQ1_CAM AND PLUS_RENDERING_ENABLED)
   TARGET_LINK_LIBRARIES(vtkInfraredTEQ1CamTest vtkPlusDataCollection vtkPlusCommon ${PLUSLIB_VTK_PREFIX}InteractionImage)
 ENDIF()
 
+#*********************** InfraredTEEV2CameraDeviceTest *****************************
+IF(PLUS_USE_INFRARED_TEEV2_CAM AND PLUS_RENDERING_ENABLED)
+  ADD_EXECUTABLE(vtkInfraredTEEV2CamTest vtkInfraredTEEV2CamTest.cxx)
+  SET_TARGET_PROPERTIES(vtkInfraredTEEV2CamTest PROPERTIES FOLDER Tests)
+  TARGET_LINK_LIBRARIES(vtkInfraredTEEV2CamTest vtkPlusDataCollection vtkPlusCommon ${PLUSLIB_VTK_PREFIX}InteractionImage)
+
+  #ADD_TEST(vtkInfraredTEEV2CamTest
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/Testing/CMakeLists.txt:516]: Remove unused code.

File: src/PlusDataCollection/UltravioletPCOUVCam/vtkUltravioletPCOUVCam.cxx [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    
+    iRet = PCO_GetRecordingState(this->cam, &RecordingState);
+    if (RecordingState)
+    {
+        iRet = PCO_SetRecordingState(this->cam, 0);
+    }
+
+ 
+    /*
+    if ((this->SetExposureTime(10000, 1)) != PLUS_SUCCESS) {
```
Comment (Sunderlandkyl) [inline src/PlusDataCollection/UltravioletPCOUVCam/vtkUltravioletPCOUVCam.cxx:141]: Remove unused code.

**Review by Sunderlandkyl — COMMENTED** [review]
Thanks for your contribution!

Couple of comments.
- Could you also rename your device classes to start with vtkPlus (vtkPlusInfraredTEEV2Cam, vtkPlusUltravioletPCOUVCam).
- Code block indentation is 2 spaces. If you use astyle, you can use this file: https://github.com/PlusToolkit/PlusLib/blob/master/.astylerc.
- Are the "FM MODIFICACION" commands tracking version changes? They should be removed.
- Could you also check that your code comments are in English?

**Comment by Sunderlandkyl** [comment]
There is currently a merge conflict. Can you rebase your changes?

**Comment by fmarcanoull** [comment]
I'm having problem with the merge.  The Cmake has the two cameras selected (PLUS_USE_INFRARED_TEEV2_CAM, PLUS_USE_ULTRAVIOLET_PCOUV_CAM), Cmake finds the SDK's, the code compiles, the PlusApp installer is generated and installed without problem, and PlusServerLauncher executes fine, but when I try to Launch any of the two cameras, there is an error, as if the devices were not set in the CMake options and compiled:

"020923_161842.037|ERROR|000.005000| Unknown device type: UltravioletPCOUVCam. Supported devices: ChRobotics, FakeTracker, GenericSerialDevice, ImageProcessor, Microchip, MmfVideo, NoiseVideo, OpenIGTLinkTracker, OpenIGTLinkVideo, SavedDataSource, UsSimulator, VirtualBufferedCapture, VirtualCapture, VirtualDeinterlacer, VirtualDiscCapture, VirtualMixer, VirtualSwitcher, VirtualVolumeReconstructor| in D:\dev\Plus-bin\PlusLib\src\PlusDataCollection\vtkPlusDeviceFactory.cxx(518)" 

I'm recompiling the code again. If the problem persist, it'll be possible I have to ask you for some help to find where I have the misconfiguration.

**Comment by fmarcanoull** [comment]
It seems to be ok now and devices are recognized. Please, tell me if the request has been sent correctly.

**Comment by fmarcanoull** [comment]
Please, tell me if there are any conflicts yet.

**Comment by Sunderlandkyl** [comment]
Yes, the conflicts are still there. When you updated your branch a couple of days ago, some commits were brought in from the master branch. GitHub won't let this PR be merged until the conflicts are resolved.

image
image

**Comment by fmarcanoull** [comment]
Those commits were made by mistake and should be discarded. All the changes I've made are in the ULL_IACTEC branch, and there is no conflicts are show in my local fork.  I'm not sure what I've to do exactly. Please, could you help me to know what command should I have to exec to fix this?  

 
image

image

**Comment by Sunderlandkyl** [comment]
I've squashed the commits manually. Thanks for your contribution!

**Comment by fmarcanoull** [comment]
Great! Thanks a lot!

**Pull request merged**